python/test_dcr_template.py

```diff
@@ -0,0 +1,629 @@
+                                           n_step=self.n_step, exposures=exposures, use_psf=use_psf)
+        self.dcrCorr.psf = dcrModel.psf
+        self.dcrCorr.psf_avg = dcrModel.psf_avg
+
+    def tearDown(self):
+        del self.dcrCorr
+
+    def test_extract_image(self):
+        # Make j and i different slightly so we can tell if the indices get swapped
+        i = self.size//2 + 1
+        j = self.size//2 - 1
+        radius = self.kernel_size//2
+        image_vals = self.dcrCorr._extract_image_vals(j, i, radius=radius)
+        input_vals = [np.ravel(self.ref_vals[f][j - radius: j + radius + 1, i - radius: i + radius + 1])
+                      for f in range(self.n_images)]
+        self.assertFloatsEqual(np.hstack(input_vals), image_vals)
+
+    def test_insert_model_vals(self):
+        i = self.size//2 + 1
+        j = self.size//2 - 1
+        radius = self.kernel_size//2
+        test_vals = np.random.random(size=(self.n_step, self.kernel_size, self.kernel_size))
+        model_ref = np.zeros((self.n_step, self.size, self.size))
+        self.dcrCorr.model = model_ref.copy()
+        weights_ref = np.zeros((self.n_step, self.size, self.size))
+        self.dcrCorr.weights = weights_ref.copy()
+        self.dcrCorr._insert_model_vals(j, i, test_vals, radius=radius)
+        psf_use = self.dcrCorr.psf_avg
+        model_ref[:, j - radius: j + radius + 1, i - radius: i + radius + 1] += test_vals*psf_use
+        weights_ref[:, j - radius: j + radius + 1, i - radius: i + radius + 1] += psf_use
+        self.assertFloatsEqual(model_ref, self.dcrCorr.model)
+        self.assertFloatsEqual(weights_ref, self.dcrCorr.weights)
+
+    def test_calculate_psf(self):
+        data_file = "test_data/calculate_psf.npy"
+        self.dcrCorr._calc_psf_model()
+        psf_size = self.dcrCorr.psf.computeKernelImage().getArray().shape[0]
+        p0 = psf_size//2 - self.kernel_size//2
+        p1 = p0 + self.kernel_size
+        psf_new = self.dcrCorr.psf.computeKernelImage().getArray()[p0: p1, p0: p1]
+        psf_ref = np.load(data_file)
+        self.assertFloatsEqual(psf_ref, psf_new)
+
+
+class RegularizationTestCase(lsst.utils.tests.TestCase):
+    def setUp(self):
+        self.kernel_size = 5
+        self.n_step = 3
+
+    def test_spatial_regularization(self):
+        data_file = "test_data/spatial_regularization.npy"
+        reg = DcrCorrection._build_regularization(x_size=self.kernel_size, y_size=self.kernel_size,
+                                                  n_step=self.n_step, spatial_regularization=True)
+        test_reg = np.load(data_file)
+        self.assertFloatsEqual(reg, test_reg)
+
+    def test_no_regularization(self):
+        # All regularization is set to False by default
+        reg = DcrCorrection._build_regularization(x_size=self.kernel_size, y_size=self.kernel_size,
+                                                  n_step=self.n_step)
+        self.assertIsNone(reg)
```

Comment:
It is very surprising to me that `_build_regularization` could ever return `None`